### ISSUE TITLE
Automatically reduce or promote dimensionality

### ISSUE THREAD

**Issue by z0u** [vpac-innovations/rsa#11]
Currently, it's difficult to combine 3D and 2D data in a query. The user needs to take the manual step of reducing the 3D data to 2D, or an exception is thrown. Some combination of the reduction solutions presented in #9 and the promotion solutions presented in #10 should be automatically used to allow various datasets to be used together.

**Comment by z0u** [comment]
Example exceptions:

``` java
org.vpac.ndg.query.QueryRuntimeException: Failed to set coordinates of filter "normaliseB". Check dimensionality.
    at org.vpac.ndg.query.FilterAdapter.invoke(FilterAdapter.java:243)
    at org.vpac.ndg.query.sampling.FilteredPixelScalar.getScalarPixel(FilteredPixelScalar.java:50)
    at org.vpac.ndg.query.sampling.FilteredPixelScalar.getPixel(FilteredPixelScalar.java:56)
    at org.vpac.ndg.query.MulBinary.kernel(MulBinary.java:52)
    at org.vpac.ndg.query.FilterAdapter.invoke(FilterAdapter.java:247)
    at org.vpac.ndg.query.sampling.FilteredPixelScalar.getScalarPixel(FilteredPixelScalar.java:50)
    at org.vpac.ndg.query.sampling.VariableBufferScalar.transfer(VariableBufferScalar.java:51)
    at org.vpac.ndg.query.TileProcessorSingle.processTile(TileProcessorSingle.java:53)
    at org.vpac.ndg.query.Query.process(Query.java:344)
    at org.vpac.ndg.query.Query.run(Query.java:304)
    at org.vpac.ndg.query.QueryRunner.run(QueryRunner.java:69)
    at org.vpac.ndg.query.QueryRunner.run(QueryRunner.java:42)
    at org.vpac.ndg.query.QueryTest.test_3D_2D_reduce(QueryTest.java:424)
    ...
Caused by: java.lang.ArrayIndexOutOfBoundsException: 2
    at org.vpac.ndg.query.math.VectorReal.set(VectorReal.java:108)
    at org.vpac.ndg.query.FilterAdapter.invoke(FilterAdapter.java:241)
    ... 35 more
```

This one was thrown when trying to use an external `.nc` file with an RSA dataset:

``` java
java.lang.ArrayIndexOutOfBoundsException: 2
       at org.vpac.ndg.query.math.VectorReal.set(VectorReal.java:108)
       at org.vpac.ndg.query.sampling.SamplerScalar.getScalarPixel(SamplerScalar.java:84)
       at org.vpac.ndg.query.sampling.SamplerScalar.getPixel(SamplerScalar.java:91)
       at org.vpac.ndg.query.AddBinary.kernel(AddBinary.java:52)
       at org.vpac.ndg.query.FilterAdapter.invoke(FilterAdapter.java:247)
```

**Comment by z0u** [comment]
This can be reproduced by trying to add an RSA dataset to one of its own tiles. For example:

Get the bounds of one tile, and then find the location of the tile in the storage pool.

``` bash
$ rsa dataset show --tiles species/25m
** Tile extents **
index   xmin    ymin    xmax    ymax
...
4,3 2500000.00  2400000.00  2625000.00  2525000.00
...
$ rsa dataset show --ncml species/25m -e 2500000.00 2400000.00 2625000.00 2525000.00 > onetile.ncml | grep location onetile.ncml
           <netcdf location="/var/lib/ndg/storagepool/species_m25/2002-01-01T00-00-00.000/Band1_tile_x4_y3.nc" section="0:4999,0:4999"/>
```

Now construct a query that uses that same dataset and the tile file:

``` xml
<?xml version="1.0" encoding="UTF-8"?>
<query xmlns="http://www.vpac.org/namespaces/rsaquery-0.2">

    <input id="2d" href="/var/lib/ndg/storagepool/species_m25/2002-01-01T00-00-00.000/Band1_tile_x4_y3.nc" />
    <input id="3d" href="rsa:species/25m" />

    <output id="outfile" >
        <grid ref="#2d" bounds="2500000.00 2400000.00 2625000.00 2525000.00" />

        <variable name="Band1" ref="#combine/output">
            <attribute name="_FillValue" value="0" />
        </variable>
    </output>

    <filter id="combine" cls="org.vpac.ndg.query.AddBinary">
        <sampler name="inputA" ref="#2d/Band1" />
        <sampler name="inputB" ref="#3d/Band1" />
    </filter>
</query>
```

In order to run this query, the FileDatasetProvider needs to be enabled:

``` diff
diff --git a/src/cmdclient/src/org/vpac/ndg/cli/Client.java b/src/cmdclient/src/org/vpac/ndg/cli/Client.java
index a8b8adb..65a0f13 100644
--- a/src/cmdclient/src/org/vpac/ndg/cli/Client.java
+++ b/src/cmdclient/src/org/vpac/ndg/cli/Client.java
@@ -165,7 +165,7 @@ public class Client {
         tileManager = (TileManager)appContext.getBean("tileManager");

         DatasetProvider dataProvider = (DatasetProvider)appContext.getBean("dataProvider");
-        ProviderRegistry.getInstance().clearProivders();
+//      ProviderRegistry.getInstance().clearProivders();
         ProviderRegistry.getInstance().addProvider(dataProvider);
         log.debug("initialisation complete");
     }
```

After rebuilding, run the query:

``` bash
$ rsa data query 2d3d.xml
ff808181466a535101466a61ff010002 0/1     0% ERROR o.v.n.c.s.l.LocalDataQuery - Task exited abnormally: 
java.lang.ArrayIndexOutOfBoundsException: 2
    at org.vpac.ndg.query.math.VectorReal.set(VectorReal.java:108) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.sampling.SamplerScalar.getScalarPixel(SamplerScalar.java:84) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.sampling.SamplerScalar.getPixel(SamplerScalar.java:91) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.AddBinary.kernel(AddBinary.java:52) ~[filter/:na]
    at org.vpac.ndg.query.FilterAdapter.invoke(FilterAdapter.java:247) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.sampling.FilteredPixelScalar.getScalarPixel(FilteredPixelScalar.java:50) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.sampling.VariableBufferScalar.transfer(VariableBufferScalar.java:51) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.TileProcessorSingle.processTile(TileProcessorSingle.java:53) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.Query.process(Query.java:344) ~[rsacli-local.jar:na]
    at org.vpac.ndg.query.Query.run(Query.java:304) ~[rsacli-local.jar:na]
    at org.vpac.ndg.cli.smadaptor.local.LocalDataQuery.execute(LocalDataQuery.java:137) ~[rsacli-local.jar:na]
    at org.vpac.ndg.cli.smadaptor.local.LocalDataQuery.access$000(LocalDataQuery.java:38) ~[rsacli-local.jar:na]
    at org.vpac.ndg.cli.smadaptor.local.LocalDataQuery$1.run(LocalDataQuery.java:113) ~[rsacli-local.jar:na]
    at java.lang.Thread.run(Thread.java:744) [na:1.7.0_55]

ERROR o.v.n.c.Client - Execution failed: Task failed to complete: 2
NGSession 2: 127.0.0.1: org.vpac.ndg.cli.Client exited with status 2
```

**Comment by z0u** [comment]
Done and merged in 4e115f08bc8b048667905d38dc7f3d9a536f4019. Filters can now specify constraints like:

``` java
    @Rank(group="in", promote=true)
    public PixelSource inputA;

    @Rank(group="in", promote=true)
    public PixelSource inputB;
```

The above would mean both inputs have to have the same number of dimensions; if not, the smaller one will be given more virtual dimensions. Demotion can be specified by the filter, too. For more details, see the `Rank` class.
